mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/resourcecluster/ResourceClustersHostManagerActor.java

```diff
@@ -328,10 +329,52 @@ private void onScaleResourceClusterRequest(ScaleResourceRequest req) {
     private void onUpgradeClusterContainersRequest(UpgradeClusterContainersRequest req) {
         log.info("Entering onScaleResourceClusterRequest: " + req);
         // [Notes] for scaling-up the request can go straight into provider to increase desire size.
-        // FOr scaling-down the decision requires getting idle hosts first.
-
-        pipe(this.resourceClusterProvider.upgradeContainerResource(req), getContext().dispatcher()).to(getSender());
+        // For scaling-down the decision requires getting idle hosts first.
+        // if enableSkuSpecUpgrade is true, first fetch the latest spec to override the sku spec during upgrade
+        // workflow.
+
+        CompletionStage<UpgradeClusterContainersResponse> upgradeFut;
+        if (req.isEnableSkuSpecUpgrade()) {
+            if (req.getResourceClusterSpec() != null) {
+                pipe(CompletableFuture.completedFuture(
+                    UpgradeClusterContainersResponse
+                        .builder()
+                        .message("Sku Spec from upgrade request is not supported. Please upgrade spec via Register "
+                            + "API first")
+                    .responseCode(ResponseCode.CLIENT_ERROR_CONFLICT)
+                    .build()),
+                    getContext().dispatcher())
+                    .to(getSender());
+                log.info("Invalid req containing custom sku spec: {}", req.getClusterId());
+                return;
+            }
+
+            upgradeFut = this.resourceClusterStorageProvider.getResourceClusterSpecWritable(req.getClusterId())
+                .thenCompose(specW -> {
+                    if (specW == null) {
+                        return CompletableFuture.completedFuture(UpgradeClusterContainersResponse.builder()
+                            .responseCode(ResponseCode.CLIENT_ERROR_NOT_FOUND)
+                            .build());
+                    }
+
+                    UpgradeClusterContainersRequest enrichedReq =
+                        req.toBuilder()
+                            .resourceClusterSpec(specW.getClusterSpec())
+                            .build();
+                    return this.resourceClusterProvider.upgradeContainerResource(enrichedReq);
+                })
```

Comment:
I think ideally we should not have this here as it increases the complexity of this endpoint.

Reply:
Temporal workflow doesn't understand how to get the spec from internal kvDal translation. So the choices are between:
1. Have workflow calling back to master endpoint to GET resource spec then apply it in the workflow.
2. Have the actor include the spec to send to workflow invoker directly  (current). I think this is cleaner and less interactions needed.